design_docs/rsyslog_new_input_output_roles_proposal.md

```diff
@@ -0,0 +1,254 @@
+# Rsyslog - Proposal for New Input and Output Role Configuration
+
+## Background
+
+The logging role defines `rsyslog_default` variable to deploy the original, all-in-one rsyslog configuration file rsyslog.conf.
+It configures to get inputs from `imjournal` and output to the local files in /var/log.
+It was originally added to provide "noop" to the logging role.
+Then, it was utilized to support for the local system logs to store into the local files combining with the openshift logging in the ovirt logging.
+In the linux-system-roles review process, some critical questions on `rsyslog_default` were raised, such as the original intention - providing "noop" is not the style ansible recommends and no other linux-system-roles do not nor will not have such a default variable.
+Considering the inputs, the functionality that `rsyslog_default` offers has to be achieved by the logging role itself instead of copying the hardcoded rsyslog.conf to the target host.
+
+This proposal will discuss about these items:
+- how to replace the `rsyslog_default` variable with the logging role input/output role configuration.
+- how to support more generic way to connect input roles and output roles.
+
+Please note that this github commit starting this doc implements the first cut of the proposal including:
+- input roles - basics input role (`imjournal`, `imuxsock`, `imudp`, `imptcp`) and files input role (`imfile`), and
+- output roles - files output role (`omfile`) and forwards output role (`omfwd`).
+Leftover tasks, other roles to be changed and eliminating `rsyslog_default`, are discussed in [To Do's](#to-dos).
+
+## Proposals
+
+### Output roles
+Each output role would be stored in a configuration file which suffix is a type name.
+
+Files and Forwards output configuration files in /etc/rsyslog.d:
+```
+  10-output-modules.conf
+  30-output.files
+  30-output.forwards
+```
+
+### Ruleset to switch the output
+Using rsyslog rulesets, we prepare all possible output combinations in the ruleset configuration file as follows.  
+Note: The ruleset name is made from the output types in the alphabetical order.
+```
+==> 50-default-rulesets.conf <==
+# Ansible managed
+
+#
+# Rules for the local system logs
+#
+ruleset(name="files") {
+  $IncludeConfig /etc/rsyslog.d/*.files
+}
+ruleset(name="forwards") {
+  $IncludeConfig /etc/rsyslog.d/*.forwards
+}
+ruleset(name="files_forwards") {
+  $IncludeConfig /etc/rsyslog.d/*.files
+  $IncludeConfig /etc/rsyslog.d/*.forwards
```

Comment:
Is this the convention that debops uses?
If not - I think it will be confusing to have rsyslog config files with different suffixes - .conf, .files, .forwards, etc. - but it can't be .conf because it will be picked up by the main rsyslog.conf `$IncludeConfig /etc/rsyslog.d/*.conf` . . . what about having separate subdirectories?  e.g. /etc/rsyslog.d/files/ - or did we already say that wasn't a good idea either?

Reply:
debops uses *.system for the system logs and *.remote for the logs sent from the remote hosts.  I extended it.

No, not considered the subdirectory in this case.  We used to have /etc/rsyslog.d/{elasticsearch,viaq}, etc.  then flattened them.  That was the only opportunity we tried the subdirectories so far.  I'd think there's no difficulties to switch suffixes to subdirectories.

Reply:
> debops uses *.system for the system logs and *.remote for the logs sent from the remote hosts. I extended it.

Ok - works for me.

> 
> No, not considered the subdirectory in this case. We used to have /etc/rsyslog.d/{elasticsearch,viaq}, etc. then flattened them. That was the only opportunity we tried the subdirectories so far. I'd think there's no difficulties to switch suffixes to subdirectories.

I think the filename suffix `*.files`, `*.remote`, etc. is fine.  I think it is an implementation detail anyway - users will typically not need to know or care about this.